### ISSUE TITLE
Error in included task in block level does not trigger always section 

### ISSUE THREAD

**Issue by betrcode** [ansible/ansible#15963]
Can't find this reported anywhere.
##### ISSUE TYPE
- Bug Report
##### ANSIBLE VERSION

```
ansible 2.0.1.0
```
##### CONFIGURATION
- No settings in ansible.cfg 
- Not using ANSIBLE_\* environment variables
##### OS / ENVIRONMENT
- Running from Ubuntu 12.04
- Managing CentOS 7
##### SUMMARY

Trying to use "- block:" with "always:" section to always put an instance back into a load balancer even when some task in the playbook has failed.

This seems to work only when tasks are inlined in the playbook, but not when they are included from another file.
##### STEPS TO REPRODUCE

Create `test-fail.yml` with contents:

```
- command: /bin/false
```

Create the `block-test.yml` playbook with contents:

```

---

- hosts: localhost

  tasks:
    - block:
      - debug: msg='i execute normally'
      - include: test-fail.yml
      - debug: msg='i never execute, cause ERROR!'
      always:
      - debug: msg='in always section'
```

Run the playbook:
`ansible-playbook block-test.yml`
##### EXPECTED RESULTS

The "in always section" message should be printed on screen.
##### ACTUAL RESULTS

Actual result:

```
 [WARNING]: provided hosts list is empty, only localhost is available

Loaded callback default of type stdout, v2.0
1 plays in block-test.yml

PLAY ***************************************************************************

TASK [setup] *******************************************************************
ESTABLISH LOCAL CONNECTION FOR USER: max
127.0.0.1 EXEC /bin/sh -c '( umask 22 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1463994869.68-81529732628427 `" && echo "` echo $HOME/.ansible/tmp/ansible-tmp-1463994869.68-81529732628427 `" )'
127.0.0.1 PUT /tmp/tmpwg3sKD TO /home/max/.ansible/tmp/ansible-tmp-1463994869.68-81529732628427/setup
127.0.0.1 EXEC /bin/sh -c 'LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/max/.ansible/tmp/ansible-tmp-1463994869.68-81529732628427/setup; rm -rf "/home/max/.ansible/tmp/ansible-tmp-1463994869.68-81529732628427/" > /dev/null 2>&1'
ok: [localhost]

TASK [debug] *******************************************************************
task path: /home/max/repos/lookup-ansible/block-test.yml:7
ok: [localhost] => {
    "msg": "i execute normally"
}

TASK [include] *****************************************************************
task path: /home/max/repos/lookup-ansible/block-test.yml:8
included: /home/max/repos/lookup-ansible/test-fail.yml for localhost

TASK [command] *****************************************************************
task path: /home/max/repos/lookup-ansible/test-fail.yml:1
ESTABLISH LOCAL CONNECTION FOR USER: max
127.0.0.1 EXEC /bin/sh -c '( umask 22 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1463994871.45-243613082978293 `" && echo "` echo $HOME/.ansible/tmp/ansible-tmp-1463994871.45-243613082978293 `" )'
127.0.0.1 PUT /tmp/tmp2_rx5D TO /home/max/.ansible/tmp/ansible-tmp-1463994871.45-243613082978293/command
127.0.0.1 EXEC /bin/sh -c 'LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/max/.ansible/tmp/ansible-tmp-1463994871.45-243613082978293/command; rm -rf "/home/max/.ansible/tmp/ansible-tmp-1463994871.45-243613082978293/" > /dev/null 2>&1'
fatal: [localhost]: FAILED! => {"changed": true, "cmd": ["/bin/false"], "delta": "0:00:00.001532", "end": "2016-05-23 11:14:31.497492", "failed": true, "invocation": {"module_args": {"_raw_params": "/bin/false", "_uses_shell": false, "chdir": null, "creates": null, "executable": null, "removes": null, "warn": true}, "module_name": "command"}, "rc": 1, "start": "2016-05-23 11:14:31.495960", "stderr": "", "stdout": "", "stdout_lines": [], "warnings": []}
        to retry, use: --limit @block-test.retry

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=1   
```
##### THE INCLUDE IS TO BLAME

If I change the playbook to not use a include, and instead inline the `- command: /bin/false`, then it works as expected. 

```

---

- hosts: localhost

  tasks:
    - block:
      - debug: msg='i execute normally'
      #- include: test-fail.yml
      - command: /bin/false
      - debug: msg='i never execute, cause ERROR!'
      always:
      - debug: msg='in always section'
```

Result:

```
PLAY ***************************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "i execute normally"
}

TASK [command] *****************************************************************
fatal: [localhost]: FAILED! => {"changed": true, "cmd": ["/bin/false"], "delta": "0:00:00.001376", "end": "2016-05-23 11:40:39.351206", "failed": true, "rc": 1, "start": "2016-05-23 11:40:39.349830", "stderr": "", "stdout": "", "stdout_lines": [], "warnings": []}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "in always section"
}

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0   

```

**Comment by betrcode** [comment]
Aha!
Just found that if I add a `rescue` section, then it starts to work!

```
---

- hosts: localhost

  tasks:
    - block:
      - debug: msg='i execute normally'
      - include: test-fail.yml
      - debug: msg='i never execute, cause ERROR!'
      rescue:
      - debug: msg='if something fails'
      always:
      - debug: msg='in always section'
```

```
PLAY ***************************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "i execute normally"
}

TASK [include] *****************************************************************
included: /home/max/repos/lookup-ansible/test-fail.yml for localhost

TASK [command] *****************************************************************
fatal: [localhost]: FAILED! => {"changed": true, "cmd": ["/bin/false"], "delta": "0:00:00.001431", "end": "2016-05-23 11:43:30.743262", "failed": true, "rc": 1, "start": "2016-05-23 11:43:30.741831", "stderr": "", "stdout": "", "stdout_lines": [], "warnings": []}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "if something fails"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "in always section"
}

PLAY RECAP *********************************************************************
localhost                  : ok=5    changed=0    unreachable=0    failed=0   
```

Still looks like a bug to me. I shouldn't have to define a rescue section if I only want a always section, right?

**Comment by berlic** [comment]
also in `ansible 2.0.2.0`

`ansible 2.1.0.0` (RC4) throws error: `ERROR! Unexpected Exception: min() arg is an empty sequence`  
here: https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/strategy/linear.py#L72

**Comment by berlic** [comment]
also if you end the `rescue` part with failed task, ansible will give the same unexpected error:

```
    - block:
        - fail: msg=fail
      rescue:
        - debug: msg=rescue
        - fail: msg=fail-in-rescue
      always:
        - debug: msg=always
```

**Comment by nfnty** [comment]
Same problem here. Traceback:

```
ERROR! Unexpected Exception: min() arg is an empty sequence
the full traceback was:

Traceback (most recent call last):
  File "/usr/bin/ansible-playbook", line 92, in <module>
    exit_code = cli.run()
  File "/usr/lib/python2.7/site-packages/ansible/cli/playbook.py", line 154, in run
    results = pbex.run()
  File "/usr/lib/python2.7/site-packages/ansible/executor/playbook_executor.py", line 142, in run
    result = self._tqm.run(play=play)
  File "/usr/lib/python2.7/site-packages/ansible/executor/task_queue_manager.py", line 256, in run
    play_return = strategy.run(iterator, play_context)
  File "/usr/lib/python2.7/site-packages/ansible/plugins/strategy/linear.py", line 174, in run
    host_tasks = self._get_next_task_lockstep(hosts_left, iterator)
  File "/usr/lib/python2.7/site-packages/ansible/plugins/strategy/linear.py", line 73, in _get_next_task_lockstep
    (s.cur_block for h, (s, t) in host_tasks_to_run
ValueError: min() arg is an empty sequence
```

**Comment by anthcourtney** [comment]
+1 to encountering this behaviour and the "ERROR! Unexpected Exception: min() arg is an empty sequence" error.

**Comment by kustodian** [comment]
The exact same thing happens if a module fails inside a block in a dependency role which has `always`.

Here is the traceback I have on 2.1.2.0 RC2, but it also fails like this on devel:

```
Traceback (most recent call last):
  File "/usr/bin/ansible-playbook", line 92, in <module>
    exit_code = cli.run()
  File "/usr/lib/python2.7/dist-packages/ansible/cli/playbook.py", line 154, in run
    results = pbex.run()
  File "/usr/lib/python2.7/dist-packages/ansible/executor/playbook_executor.py", line 146, in run
    result = self._tqm.run(play=play)
  File "/usr/lib/python2.7/dist-packages/ansible/executor/task_queue_manager.py", line 264, in run
    play_return = strategy.run(iterator, play_context)
  File "/usr/lib/python2.7/dist-packages/ansible/plugins/strategy/linear.py", line 174, in run
    host_tasks = self._get_next_task_lockstep(hosts_left, iterator)
  File "/usr/lib/python2.7/dist-packages/ansible/plugins/strategy/linear.py", line 73, in _get_next_task_lockstep
    (s.cur_block for h, (s, t) in host_tasks_to_run
ValueError: min() arg is an empty sequence
```

**Comment by kustodian** [comment]
I would like to note, that this doesn't happen only if you have an include in the block element, it also fails if the whole block is in an included file, for example:

```
# main.yml
- include: block.yml
```

```
# block.yml
- block:
    - fail: msg=Fail
  rescue:
    - debug: msg=Whatever
```

So if you use a lot of includes, this can happen to a lot of people.

**Comment by kustodian** [comment]
Here is a repro which I have, it's a little different:

```
# test_blocks.yml
---
- hosts: localhost
  connection: local
  tasks:
    - include: blocks.yml
```

```
# blocks.yml
- block:
    - fail: msg='This will fail'
  always:
    - debug: msg='This will execute and stop'

- debug: msg="This shouldn't be executed"
```

When I run this command against latest `stable-2.1` branch:

```
ansible-playbook -i localhost, test_blocks.yml -vvv
```

I get this output:

```
PLAYBOOK: test_blocks.yml ******************************************************
1 plays in test_blocks.yml

PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: user
<localhost> EXEC /bin/sh -c 'LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python && sleep 0'
ok: [localhost]

TASK [fail] ********************************************************************
task path: /home/user/git/ansible-server-deploy/ansible_playbooks/blocks.yml:2
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_args": {"msg": "This will fail"}, "module_name": "fail"}, "msg": "This will fail"}
ERROR! Unexpected Exception: min() arg is an empty sequence
the full traceback was:

Traceback (most recent call last):
  File "/home/user/git/ansible-devel/bin/ansible-playbook", line 92, in <module>
    exit_code = cli.run()
  File "/home/user/git/ansible-devel/lib/ansible/cli/playbook.py", line 154, in run
    results = pbex.run()
  File "/home/user/git/ansible-devel/lib/ansible/executor/playbook_executor.py", line 146, in run
    result = self._tqm.run(play=play)
  File "/home/user/git/ansible-devel/lib/ansible/executor/task_queue_manager.py", line 264, in run
    play_return = strategy.run(iterator, play_context)
  File "/home/user/git/ansible-devel/lib/ansible/plugins/strategy/linear.py", line 174, in run
    host_tasks = self._get_next_task_lockstep(hosts_left, iterator)
  File "/home/user/git/ansible-devel/lib/ansible/plugins/strategy/linear.py", line 73, in _get_next_task_lockstep
    (s.cur_block for h, (s, t) in host_tasks_to_run
ValueError: min() arg is an empty sequence
```

**Comment by jimi-c** [comment]
# Closing This Ticket

Hi!

We believe the above commit should resolve this problem for you. This will also be included in the next release. This was also cherry-picked to the `stable-2.1` branch, and will be included in 2.1.2 (it will be present in RC4 on).

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
